Feature/dpl 016 stop parsing unconsolidated data from specified centres

sdjmchattie opened the pull request:
Closes #346 

Changes proposed in this pull request:

* Added a new file state for files that shouldn't be processed.
* Set that state if the centre defines that it shouldn't process unconsolidated files and the file is of that type.
* Generate an error when the file has that state, but still make a backup of the file in the errors directory so that the file is not processed again in the future.
* Update tests for the above and add new ones to test the new functionality.

pjvv commented on crawler/file_processing.py:
```diff
@@ -386,6 +395,14 @@ def set_state_for_file(self) -> CentreFileState:
 
+        # check for this being an unconsolidated samples file where the centre doesn't support those
+        elif (
+                CENTRES_KEY_SKIP_UNCONSOLIDATED_FILES in centre and
```
Could we use `centre.get(CENTRES_KEY_SKIP_UNCONSOLIDATED_FILES, False) and self.is_unconsolidated_surveillance_file() ` instead?

pjvv commented on crawler/file_processing.py:
```diff
@@ -439,6 +456,22 @@ def process_samples(self, add_to_dart: bool) -> None:
+    def log_unprocessed(self) -> None:
+        """Log the file as unprocessed and ensure it won't be processed in future.
+
+        In this implementation, we assume the reason the file shouldn't be processed is that
```
Is it more accurate to say that "we (Sanger) do not want to support unconsolidated files from the centre anymore"? I might have missed some context.

pjvv commented on crawler/file_processing.py:
```diff
@@ -439,6 +456,22 @@ def process_samples(self, add_to_dart: bool) -> None:
+        self.logging_collection.add_error(
+            "TYPE 34",
+            f"File '{self.file_name}' is not being processed because unconsolidated "
+            "sample files are unsupported by this centre.")
```
Something along the lines of "because unconsolidated sample files from this centre and unsupported" instead?

pjvv commented on crawler/helpers/logging_helpers.py:
```diff
@@ -388,12 +388,24 @@ def __init__(self):
         self.error_level = ErrorLevel.ERROR
         self.message = (
-            "{self.error_level.name}: Priority samples where the DART database inserts have failed for some plates. "
+            f"{self.error_level.name}: Priority samples where the DART database inserts have failed for some plates. "
```
Good spot, thanks.

pjvv commented on crawler/helpers/logging_helpers.py:
```diff
@@ -388,12 +388,24 @@ def __init__(self):
+        self.type_str = "TYPE 34"
+        self.error_level = ErrorLevel.ERROR
+        self.message = (
+            f"{self.error_level.name}: Unconsolidated samples file provided by a centre that does not support them. "
```
Message here might need changing too if previous comments are valid.

pjvv commented on crawler/helpers/logging_helpers.py:
```diff
@@ -388,12 +388,24 @@ def __init__(self):
+            f"{self.error_level.name}: Unconsolidated samples file provided by a centre that does not support them. "
+            f"({self.type_str})"
+        )
+        self.short_display_description = "File provided by unsupported centre"
```
Ditto.

pjvv commented on tests/test_file_processing.py:
```diff
@@ -338,7 +368,7 @@ def create_checksum_files_for(filepath, filename, checksums, timestamp):
 def test_checksum_not_match(config, tmpdir):
     with patch.dict(config.CENTRES[0], {"backups_folder": tmpdir.realpath()}):
-        tmpdir.mkdir("successes")
+        tmpdir.mkdir(SUCCESSES_DIR)
```
Thanks for this improvement.

pjvv commented on tests/test_main.py:
```diff
@@ -38,35 +38,42 @@ def test_run(mongo_database, testing_files_for_process, pyodbc_conn):
-    assert source_plates_collection.count_documents({}) == NUMBER_SOURCE_PLATES
-    assert source_plates_collection.count_documents({"barcode": "AP123"}) == 1
+    assert source_plates_collection.count_documents({}) == NUMBER_ACCEPTED_SOURCE_PLATES
+    # Centres that don't process unconsolidated files
```
Comment might be misleading too if previous comments are valid.

sdjmchattie commented:
@pjvv Thanks for your prior review.  I've updated everything according to your comments and I've added a couple more tests.  I thought that would cause the codecov bot to say we have 100% coverage now, but it still thinks I've removed coverage.  I can't explain why it's saying that.  I also not did all the lifting and static analysis checks and fixed those too.

pjvv commented on Pipfile:
```diff
@@ -25,6 +25,7 @@ more-itertools = "~=8.6"
 pandas = "~=1.1"
 sqlalchemy = "~=1.3"
 pymysql = "~=1.0"
+types-python-dateutil = "*"
```
This should be moved to `dev-packages`. Easiest is to `pipenv uninstall` then `pipenv install --dev`.

sdjmchattie replied:
Sorry yes of course 👍 will update
